docs/design/services/dynamodb/high-level-library/prototype/option-1/sync/Prototype.java

```diff
@@ -0,0 +1,1026 @@
+     *
+     * This is equivalent to {@code ItemAttributeValue.from(object).convertFromJavaType(converter)}.
+     */
+    static ItemAttributeValue from(Object object, ItemAttributeValueConverter<?> converter);
+
+    /**
+     * Create an {@link ItemAttributeValue} that represents the DynamoDB-specific null type.
+     */
+    static ItemAttributeValue nullValue();
+
+    /**
+     * Convert this item attribute value into the requested Java type.
+     *
+     * This uses the {@link ItemAttributeValueConverter} configured on this type via
+     * {@link #from(Object, ItemAttributeValueConverter)} or {@link #convertFromJavaType(ItemAttributeValueConverter)}.
+     */
+    <T> T as(Class<T> type);
+
+    /**
+     * Retrieve the {@link ItemAttributeValueType} of this value.
+     */
+    ItemAttributeValueType type();
+
+    /**
+     * The {@code is*} methods can be used to check the underlying DynamoDB-specific type of the attribute value.
+     *
+     * If the type isn't known (eg. because it was created via {@link ItemAttributeValue#from(Object)}), {@link #isJavaType()}
+     * will return true. Such types will be converted into DynamoDB-specific types by the document client before they are
+     * persisted.
+     */
+
+    boolean isItem();
+    boolean isString();
+    boolean isNumber();
+    boolean isBytes();
+    boolean isBoolean();
+    boolean isListOfStrings();
+    boolean isListOfNumbers();
+    boolean isListOfBytes();
+    boolean isListOfAttributeValues();
+    boolean isNull();
+    boolean isJavaType();
+
+    /**
+     * The {@code as*} methods can be used to retrieve this value without the overhead of type conversion of {@link #as(Class)}.
+     *
+     * An exception will be thrown from these methods if the requested type does not match the actual underlying type. When
+     * the type isn't know, the {@code is*} or {@link #type()} methods can be used to query the underlying type before
+     * invoking these {@code as*} methods.
+     */
+    Item asItem();
+    String asString();
+    BigDecimal asNumber();
+    SdkBytes asBytes();
+    Boolean asBoolean();
+    List<String> asListOfStrings();
+    List<BigDecimal> asListOfNumbers();
+    List<SdkBytes> asListOfBytes();
+    List<ItemAttributeValue> asListOfAttributeValues();
+    Object asJavaType();
+
+    /**
+     * Convert this attribute value from a {@link ItemAttributeValueType#JAVA_TYPE} to a type that can be persisted in DynamoDB.
+     *
+     * This will throw an exception if {@link #isJavaType()} is false.
+     */
+    ItemAttributeValue convertFromJavaType(ItemAttributeValueConverter<?> converter);
+}
+
+/**
+ * The schema for a specific item. This describes the item's structure and which attributes it contains.
+ *
+ * This is mostly an implementation detail, and can be ignored except by developers interested in creating
+ * {@link ItemAttributeValueConverter}.
+ */
+@ThreadSafe
+public interface ItemSchema {
```

Comment:
How will polymorphism be supported? Will item schemas nest?

Reply:
Inheritance isn't implemented in this prototype, but if it were... in my mind, an `ItemSchema` is the runtime equivalent of a class definition.

Class Definitions:
```Java
@ChildTypes(Dog.class)
public class Animal {
    @Index(PARTITION_KEY)
    @Attribute
    private UUID id;

    @Attribute
    private Instant birthDate;

    // getters/setters
}

@Converter(MyDogConverter.class)
public class Dog extends Animal {
    @Attribute
    @Converter(BreedConverter.class)
    private String breed;

    // getters/setters
}
```

Item Schemas:
```
{
    javaType: Animal.class,
    childTypes: [ Dog.class ],
    attributeSchemas: {
        id: {
            indexType: PARTITION_KEY,
            javaType: UUID.class
        },
        birthDate: {
            javaType: Instant.class
        }
    }
}

{
    javaType: Dog.class,
    converter: MyDogConverter.class,
    attributeSchemas: {
        breed: {
            javaType: String.class
            converter: BreedConverter.class
        }
    }
}
```